codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -69,60 +69,113 @@
     private final Http2Settings initialSettings;
     private ChannelFutureListener closeListener;
     private BaseDecoder byteDecoder;
-    private long gracefulShutdownTimeoutMillis = DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS;
+    private long gracefulShutdownTimeoutMillis;
 
-    public Http2ConnectionHandler(boolean server) {
-        this(server, true);
+    public static final class Builder extends BuilderBase<Http2ConnectionHandler, Builder> {
+        @Override
+        public Http2ConnectionHandler build(Http2ConnectionDecoder decoder,
+                                            Http2ConnectionEncoder encoder) {
+            Http2ConnectionHandler handler = new Http2ConnectionHandler(decoder, encoder, initialSettings());
+            setPostBuildOptions(handler);
+            return handler;
+        }
     }
 
-    public Http2ConnectionHandler(boolean server, boolean validateHeaders) {
-        this(new DefaultHttp2Connection(server), validateHeaders);
-    }
+    public abstract static class BuilderBase<T extends Http2ConnectionHandler, B extends BuilderBase<T, B>> {
+        private Http2Settings initialSettings;
+        private Http2FrameListener listener;
+        private boolean validateHttp2Headers = true;
+        private boolean server = true;
+        private Http2FrameLogger frameLogger;
+        private long gracefulShutdownTimeoutMillis = DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS;
 
-    public Http2ConnectionHandler(Http2Connection connection) {
-        this(connection, true);
-    }
+        @SuppressWarnings("unchecked")
+        public B frameListener(Http2FrameListener listener) {
+            this.listener = listener;
+            return (B) this;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, boolean validateHeaders) {
-        this(connection, new DefaultHttp2FrameReader(validateHeaders), new DefaultHttp2FrameWriter());
-    }
+        public final Http2FrameListener frameListener() {
+            return listener;
+        }
 
-    public Http2ConnectionHandler(Http2Connection connection, Http2FrameReader frameReader,
-                                  Http2FrameWriter frameWriter) {
-        initialSettings = null;
-        encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, frameReader);
-    }
+        @SuppressWarnings("unchecked")
+        public B validateHeaders(boolean validate) {
```

Comment:
Thought about this....it is awkward to have this in the base class. I moved the "set both" method into the inheriting class where it actually does something different.